engine-types/src/parameters/engine.rs

```diff
@@ -177,16 +177,79 @@ pub struct ResultLog {
     pub data: Vec<u8>,
 }
 
+/// EVM error king of transaction status
+#[derive(Debug, Clone, BorshSerialize, BorshDeserialize, PartialEq, Eq)]
+#[cfg_attr(feature = "impl-serde", derive(Serialize, Deserialize))]
+pub enum EvmErrorKind {
+    /// Trying to pop from an empty stack.
+    StackUnderflow,
+    /// Trying to push into a stack over stack limit.
+    StackOverflow,
+    /// Jump destination is invalid.
+    InvalidJump,
+    /// An opcode accesses memory region, but the region is invalid.
+    InvalidRange,
+    /// Encountered the designated invalid opcode.
+    DesignatedInvalid,
+    /// Call stack is too deep (runtime).
+    CallTooDeep,
+    /// Create opcode encountered collision (runtime).
+    CreateCollision,
+    /// Create init code exceeds limit (runtime).
+    CreateContractLimit,
+    /// Invalid opcode during execution or starting byte is 0xef. See [EIP-3541](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-3541.md).
+    InvalidCode(u8),
+    /// An opcode accesses external information, but the request is off offset
+    /// limit (runtime).
+    OutOfOffset,
+    /// Execution runs out of gas (runtime).
+    OutOfGas,
+    /// Not enough fund to start the execution (runtime).
+    OutOfFund,
+    /// PC underflowed (unused).
+    #[allow(clippy::upper_case_acronyms)]
+    PCUnderflow,
+    /// Attempt to create an empty account (runtime, unused).
+    CreateEmpty,
+    /// Other normal errors.
+    Other(crate::Cow<'static, str>),
+    /// Nonce reached maximum value of 2^64-1
+    MaxNonce,
+    /// `usize` casting overflow
+    UsizeOverflow,
+}
+
+impl AsRef<[u8]> for EvmErrorKind {
+    fn as_ref(&self) -> &[u8] {
+        match self {
+            Self::StackUnderflow => b"STACK_UNDERFLOW",
+            Self::StackOverflow => b"STACK_OVERFLOW",
+            Self::InvalidJump => b"INVALID_JUMP",
+            Self::InvalidRange => b"INVALID_RANGE",
+            Self::DesignatedInvalid => b"DESIGNATED_INVALID",
+            Self::CallTooDeep => b"CALL_TOO_DEEP",
+            Self::CreateCollision => b"CREATE_COLLISION",
+            Self::CreateContractLimit => b"CREATE_CONTRACT_LIMIT",
+            Self::InvalidCode(_) => b"INVALID_CODE",
+            Self::OutOfOffset => b"OUT_OF_OFFSET",
+            Self::OutOfGas => b"OUT_OF_GAS",
+            Self::OutOfFund => b"OUT_OF_FUND",
+            Self::PCUnderflow => b"PC_UNDERFLOW",
+            Self::CreateEmpty => b"CREATE_EMPTY",
+            Self::MaxNonce => b"MAX_NONCE",
+            Self::UsizeOverflow => b"USIZE_OVERFLOW",
+            Self::Other(e) => e.as_bytes(),
+        }
+    }
+}
+
 /// The status of a transaction.
 #[derive(Debug, Clone, BorshSerialize, BorshDeserialize, PartialEq, Eq)]
 #[cfg_attr(feature = "impl-serde", derive(Serialize, Deserialize))]
 pub enum TransactionStatus {
     Succeed(Vec<u8>),
     Revert(Vec<u8>),
-    OutOfGas,
-    OutOfFund,
-    OutOfOffset,
-    CallTooDeep,
+    Error(EvmErrorKind),
```

Comment:
This is a breaking change in the binary (borsh) format of `TransactionStatus`. For example, previously `TransactionStatus::OutOfGas` serialized as `0x02`, but now `TransactionStatus::Error(EvmErrorKind::OutOfGas)` will serialize as `0x020a`.

I think the safest thing to do is make this change backwards compatible by retaining the previous variants of `TransactionStatus` and adding `Error` as a new variant. Then the new `EvmErrorKind` should skip `OutOfGas`, `OutOfFund`, `OutOfOffset` and `CallTooDeep` so that we do not have duplicate representations of the same error.

New tests should be added checking that we catch the new error kinds that will appear in the new `TransactionStatus` variant.

Reply:
First suggestion - I agree.

About tests, for me it's a bit unclear.
Do you mean all variants of `EmvErrorKind`?

For example: `test_contract_create_too_large` cover: `TransactionStatus::Error(EvmErrorKind::CreateContractLimit)`

Reply:
For example it would be good to add a test that covers `TransactionStatus::Error(EvmErrorKind::StackUnderflow)`.

Reply:
Am I clear, that your point also to check that borsh serialization matches the expected results? As currently there are no tests for that.

Reply:
Yes, it would also be a good idea to have regression tests for the borsh serialization. I think it makes sense to test the serialization of `SubmitResult` which contains `TransactionStatus`.